### PR TITLE
adapter,interchange,parser: Support avro sink comments

### PR DESCRIPTION
**Pull request by moulimukherjee**
Added implementation to 
* automatically add comments to generated avro schemas from materialized comments (via `COMMENT ON`)
* allow user to specify documentation in the sql

Currently, `COMMENT ON` (in materialize) does not support adding comments for a column in a custom type and this PR also skips that. I will do a follow up to add support for that https://github.com/MaterializeInc/materialize/issues/22210.

### Motivation
Partially fixes https://github.com/MaterializeInc/materialize/issues/21557, does not allow comments on columns in a custom type yet.

### Tips for reviewer
Added inline comments.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
    - https://github.com/MaterializeInc/materialize/pull/21564
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - I will add user facing documentation as a follow up

cc: @benesch @dseisun-materialize

### CODE REVIEW

File: doc/developer/design/20230903_avro_doc.md
Comment (moulimukherjee) [inline doc/developer/design/20230903_avro_doc.md:117]: Updated the implementatoin to keep it consistent with the naming resolution for the type below `[[db.]schema.]object` and with the `COMMENT ON` sql. Also, sinks can't be queried and they don't have a relation desc.

File: src/sql-parser/src/ast/defs/ddl.rs
Comment (moulimukherjee) [inline src/sql-parser/src/ast/defs/ddl.rs:161]: OOC will it be a good idea or a bad idea to have something like T::SubItemName which can automatically resolve to a column name?

**Comment by umanwizard** [comment]
> Currently, COMMENT ON does not support adding comments for a column in a custom type and this PR also skips that

FWIW, this works fine in Postgres:

```
brennan=> create type t as (f1 text);
CREATE TYPE
brennan=> comment on column t.f1 is 'hello, world!';
COMMENT
```

**Comment by moulimukherjee** [comment]
@umanwizard Yeah. It's just not implemented yet in materialize. I will do a follow up to add both `comment on` support and `DOC ON COLUMN` support for them. This PR was becoming a bit too big and they'll have overlapping logic.

File: src/sql-parser/src/ast/defs/ddl.rs
Comment (mjibson) [reply src/sql-parser/src/ast/defs/ddl.rs:161]: I think that'd be a great idea! Might make a good followup PR so that `CommentObjectType::Column` can get the same treatment.

Comment (moulimukherjee) [reply src/sql-parser/src/ast/defs/ddl.rs:161]: https://materializeinc.slack.com/archives/C02FWJ94HME/p1696540204393639

File: src/sql-parser/src/parser.rs
Comment (mjibson) [inline src/sql-parser/src/parser.rs:2100]: This differs from https://github.com/MaterializeInc/materialize/blob/8b47d4c7df7deeb6fe8645262f59fb4a7d2ce60f/src/sql-parser/src/parser.rs#L7764 which calls parse_identifiers. I thiiiink the one in this PR is correct because that's what AstInfo expects. @ParkMyCar fyi: comment on column should probably be taught to call this new function once it supports AstInfo.

File: src/sql/src/plan/statement/ddl.rs
Comment (mjibson) [inline src/sql/src/plan/statement/ddl.rs:2150]: I think this logic belongs in `purify_statement`, then it doesn't need to be mut here.

File: src/sql/src/plan/statement/ddl.rs
Comment (mjibson) [inline src/sql/src/plan/statement/ddl.rs:2464]: Can you open an issue about this so we don't forget? I think keeping everything in the macro is good, so we should eventually add enum support to it.

File: src/sql-parser/src/ast/defs/ddl.rs
Comment (moulimukherjee) [reply src/sql-parser/src/ast/defs/ddl.rs:161]: Cool, I will try this as a follow up 👍

File: src/sql/src/plan/statement/ddl.rs
Comment (moulimukherjee) [reply src/sql/src/plan/statement/ddl.rs:2464]: Sure 👍. I did not find any other similar use cases, so this could be an outlier.

File: src/sql/src/plan/statement/ddl.rs
Comment (moulimukherjee) [reply src/sql/src/plan/statement/ddl.rs:2150]: Ah, let me update.

File: src/sql/src/plan/statement/ddl.rs
Comment (moulimukherjee) [reply src/sql/src/plan/statement/ddl.rs:2464]: filed https://github.com/MaterializeInc/materialize/issues/22213, will add it in the comment as well.

File: src/sql/src/plan/statement/ddl.rs
Comment (moulimukherjee) [reply src/sql/src/plan/statement/ddl.rs:2150]: I just looked at the code. The `purify_statement` is resolving external dependencies and returns vector with `CreateSubsourceStatement`, which is not applicable here since there are no external dependencies and no subsources. I could return an empty vector, but it does not seem well suited?

File: src/interchange/src/json.rs
Comment (umanwizard) [inline src/interchange/src/json.rs:261]: We should be able to avoid all these clones by passing a reference here (and in `build_row_schema_fields`) rather than an owned object.

File: src/interchange/src/json.rs
Comment (umanwizard) [inline src/interchange/src/json.rs:434]:
```suggestion
                .expect("`field` initialized to JSON object above")
```

File: src/interchange/src/json.rs
Comment (umanwizard) [inline src/interchange/src/json.rs:446]:
```suggestion
                .expect("`field` initialized to JSON object above")
```

**Review by umanwizard — COMMENTED** [review]
Changes to Avro schema generation look fine to me. Someone from Adapter should review the rest.

I think we could use an end-to-end test that verifies that other Avro-speaking tools (e.g. the official Java or Python Avro libraries) can properly interpret these schemas and see the docs on them. But I wouldn't think that should block this PR as we have nothing like that currently. @philip-stoev or @def- , what do you think?

File: src/interchange/src/json.rs
Comment (moulimukherjee) [reply src/interchange/src/json.rs:457]: Sure 👍

File: src/interchange/src/json.rs
Comment (moulimukherjee) [reply src/interchange/src/json.rs:261]: I had started with that, but the `item_id` inside it needs to be updated for nested avro records.

Comment (moulimukherjee) [reply src/interchange/src/json.rs:261]: I guess if I extract out the `item_id` as a separate method parameter that would work.

**Comment by def-** [comment]
> I think we could use an end-to-end test that verifies that other Avro-speaking tools (e.g. the official Java or Python Avro libraries) can properly interpret these schemas and see the docs on them. But I wouldn't think that should block this PR as we have nothing like that currently. @philip-stoev or @def- , what do you think?

Makes sense to have that. I'll try implementing that test on Friday, probably using Python Avro lib, since that's easiest from our side. I'd prefer if you can wait for the test.

File: src/interchange/src/json.rs
Comment (moulimukherjee) [reply src/interchange/src/json.rs]: Updated!

File: src/interchange/src/json.rs
Comment (moulimukherjee) [reply src/interchange/src/json.rs]: Updated to pass reference `&SchemaOptions` now, removed this `Rc` and all the clones.

File: src/interchange/src/json.rs
Comment (moulimukherjee) [reply src/interchange/src/json.rs]: done

File: src/sql-parser/src/parser.rs
Comment (moulimukherjee) [reply src/sql-parser/src/parser.rs]: done

File: src/sql/src/plan/statement/ddl.rs
Comment (mjibson) [reply src/sql/src/plan/statement/ddl.rs:2150]: Agree; my mistake. I hadn't looked at that function deeply and didn't realize it's only called in one place. LGTM.

**Review by mjibson — APPROVED** [review]
Adapter parts lgtm.

**Review by benesch — COMMENTED** [review]
I'm planning to take a look at this tomorrow as well! Might slip into the weekend but will get you a review by Monday morning at the latest.

**Comment by philip-stoev** [comment]
> Makes sense to have that. I'll try implementing that test on Friday, probably using Python Avro lib, since that's easiest from our side. I'd prefer if you can wait for the test.

Instead of bringing another tool into the mix, can we simply issue a GET request against the schema service and fish for the required parts in the response?

File: test/testdrive/kafka-avro-sinks-doc-comments.td
Comment (philip-stoev) [inline test/testdrive/kafka-avro-sinks-doc-comments.td:37]: @def-  given the extra syntax that is being added, I think a platform check would be nice to have here. Or, just have the existing sinks.py put some comments and then use `SHOW CREATE` to check for them in `validate()`

Comment (def-) [reply test/testdrive/kafka-avro-sinks-doc-comments.td:37]: Yes, already on my list.

**Comment by def-** [comment]
Should we have user-facing documentation for the new feature or is that planned later?

File: src/interchange/src/json.rs
Comment (def-) [inline src/interchange/src/json.rs:81]: Uncovered

File: src/sql-parser/src/parser.rs
Comment (def-) [inline src/sql-parser/src/parser.rs:2107]: Uncovered

**Review by def- — COMMENTED** [review]
Only 2 interesting uncovered spots based on coverage report: https://buildkite.com/materialize/coverage/builds/259#_

**Review by def- — CHANGES_REQUESTED** [review]
Without enable_comment calling `COMMENT ON` calls reference a wrong issue: "ERROR:  COMMENT ON not yet supported, see https://github.com/MaterializeInc/materialize/issues/20218 for more details". That issue is about webhooks' launchdarkly flag, not about `COMMENT ON`

**Review by def- — CHANGES_REQUESTED** [review]
Even with `MZ_UNSAFE_MODE=1 MZ_ALL_FEATURES=1` comment on is not yet enabled. In src/sql/src/session/vars.rs add `enable_comment` to `feature_flags!`.

**Comment by umanwizard** [comment]
@philip-stoev The reason I recommended using the official Avro library was to make sure the schema can still be parsed, i.e. that we are putting the `doc` fields in the right place. It's not too hard to do that in Python, just something like this:

```
from avro.schema import parse

schema_str = ... # get from HTTP request to CSR

schema = parse(schema_str)
fields_with_docs = [(field.name, field.doc) for field in schema.fields if field.doc]

print(fields_with_docs)```

But, maybe I'm being too paranoid.

File: src/sql/src/session/vars.rs
Comment (moulimukherjee) [inline src/sql/src/session/vars.rs:1716]: cc @ParkMyCar @def- Made this into a feature flag

**Comment by def-** [comment]
pg_description does not show the comments (or even that the column exists!):
```
materialize=> create table t1 (c int);
CREATE TABLE
materialize=> comment on table t1 is 'foobar';
COMMENT
materialize=> comment on column t1.c is 'foobar2';
COMMENT
materialize=> SELECT * FROM pg_description where objoid = 't1'::regclass;
 objoid | classoid | objsubid | description
--------+----------+----------+-------------
  20631 |          |        0 |
(1 row)
```
Meanwhile in Pg:
```
deen=> SELECT * FROM pg_description where objoid = 't1'::regclass;
 objoid | classoid | objsubid | description
--------+----------+----------+-------------
  24739 |     1259 |        0 | foobar
  24739 |     1259 |        1 | foobar2
(2 rows)
```

File: src/interchange/src/json.rs
Comment (moulimukherjee) [reply src/interchange/src/json.rs:81]: This is the `Debug` implementation only applicable when this object will be logged and I am guessing we don't log this anywhere, which is probably fine.

**Comment by moulimukherjee** [comment]
>pg_description does not show the comments (or even that the column exists!):

@def- That's related to https://github.com/MaterializeInc/materialize/issues/21451? This PR is probably not the right place to fix existing issues with `COMMENT ON` and should be addressed separately.

File: src/sql-parser/src/parser.rs
Comment (def-) [reply src/sql-parser/src/parser.rs:2107]: They should be but it ran into a timeout and thus no coverage file was created. Sorry for the false alarm.

Comment (moulimukherjee) [reply src/sql-parser/src/parser.rs:2107]: No worries 👍

**Comment by moulimukherjee** [comment]
> The reason I recommended using the official Avro library was to make sure the schema can still be parsed

@umanwizard @philip-stoev  I would +1 this. It would be nice to test that the schema is actually a valid avro schema. I was using an online tool for development purposes (And I _had_ created bad avro schema during iteration which I fixed later)

File: src/sql/src/session/vars.rs
Comment (moulimukherjee) [reply src/sql/src/session/vars.rs:1716]: Actually let me separate that out in a different PR, it's not exactly related to the changes here and there are more tests failing.

Comment (moulimukherjee) [reply src/sql/src/session/vars.rs:1716]: WIP https://github.com/MaterializeInc/materialize/pull/22242

**Comment by moulimukherjee** [comment]
> Even with MZ_UNSAFE_MODE=1 MZ_ALL_FEATURES=1 comment on is not yet enabled. In src/sql/src/session/vars.rs add enable_comment to feature_flags!.

@def- Fix here https://github.com/MaterializeInc/materialize/pull/22242

File: src/sql-parser/src/ast/defs/ddl.rs
Comment (moulimukherjee) [reply src/sql-parser/src/ast/defs/ddl.rs:161]: Addressed in https://github.com/MaterializeInc/materialize/pull/22244

**Comment by def-** [comment]
> > The reason I recommended using the official Avro library was to make sure the schema can still be parsed
> 
> @umanwizard @philip-stoev I would +1 this. It would be nice to test that the schema is actually a valid avro schema. I was using an [online validator](http://avro.tarantool.org/#) for development purposes (And I _had_ created bad avro schema during iteration which I fixed later)

That actually sounds like testdrive's `schema-registry-verify` should maybe be made smarter to find avro schema problems?

**Comment by moulimukherjee** [comment]
>That actually sounds like testdrive's schema-registry-verify should maybe be made smarter to find avro schema problems? What was the mistake and error?

@def- Yeah, if `schema-registry-verify` could validate it's a valid avro that would be nice. This was during the `NULL DEFAULTS` implementation. I had mistakenly set the default to `"null"` string instead of `null` and could publish the schema, which was wrong for a nullable `long`.

There's another class of errors which `schema-registry-verify` cannot properly verify, like is the documentation correctly set at the proper attribute name. If say the comments got added on `"docs"` instead of `"doc"`, it would be a valid schema since arbitrary metadata is allowed even if it's not setting the correct intended field here.

**Comment by def-** [comment]
If there is a comment on a column, can I create a sink without exporting that comment? I'd have expected `DOC ON COLUMN t.c1 = NULL` to work, but it fails with `cannot use value as string`.

**Comment by moulimukherjee** [comment]
> If there is a comment on a column, can I create a sink without exporting that comment?

Ah, currently no. Let me look into that.

**Comment by def-** [comment]
I added some tests:

- platform-checks: Explicit sink comment check (currently fails, @moulimukherjee this one is confusing me, the value doc fields seem to be missing, even without restarts)
- platform-checks: Added comments to Identifiers check
- parallel-workload: Enabled COMMENT ON
- testdrive: Added failure case on NULL and escaping
- sqlsmith: will come separately in https://github.com/MaterializeInc/sqlsmith/pull/3

**Comment by moulimukherjee** [comment]
Moving to draft to look into the tests

**Comment by def-** [comment]
The failure is:
```
Running validate() from <materialize.checks.sink.SinkComments object at 0x7f1d4a52bad0>
$ docker compose exec -T testdrive testdrive --kafka-addr=kafka:9092 --schema-registry-url=http://schema-registry:8081 --materialize-url=postgres://materialize@materialized:6875 --materialize-internal-url=postgres://materialize@materialized:6877 --variable-length-row-encoding --aws-endpoint=http://localstack:4566 --no-reset --default-timeout=300s --seed=1 --persist-blob-url=file:///mzdata/persist/blob --persist-consensus-url=postgres://root@materialized:26257?options=--search_path=consensus --var=replicas=1 --var=default-replica-size=4-4 --var=default-storage-size=4-1 --source=/home/deen/git/materialize/misc/python/materialize/checks/sink.py:691
Verifying contents of latest schema for subject "sink-sink-comments1-key" in the schema registry...
Verifying contents of latest schema for subject "sink-sink-comments2-key" in the schema registry...
Verifying contents of latest schema for subject "sink-sink-comments3-key" in the schema registry...
Verifying contents of latest schema for subject "sink-sink-comments1-value" in the schema registry...
^^^ +++
11:1: error: schema did not match
expected:
{"type":"record","name":"envelope","doc":"comment on view sink_source_comments_view","fields":[{"name":"before","type":["null",{"type":"record","name":"row","fields":[{"name":"l_k","type":"string"},{"name":"l_v1","type":["null","string"],"default":null,"doc":"doc on l_v1"},{"name":"l_v2","type":["null","long"],"default":null,"doc":"value doc on l_v1"},{"name":"c","type":"long"}]}],"default":null},{"name":"after","type":["null","row"],"default":null}]}

actual:
{"type":"record","name":"envelope","doc":"comment on view sink_source_comments_view","fields":[{"name":"before","type":["null",{"type":"record","name":"row","fields":[{"name":"l_k","type":"string"},{"name":"l_v1","type":["null","string"],"default":null},{"name":"l_v2","type":["null","long"],"default":null},{"name":"c","type":"long"}]}],"default":null},{"name":"after","type":["null","row"],"default":null}]}
     |
  10 |
  11 | $ schema-registry-verify schema-type=avro subject=sink-sink-comments1-value
     | ^
+++ !!! Error Report
```
Maybe I'm holding it wrong?

**Comment by moulimukherjee** [comment]
@def- I had run into a similar issue locally because for column/fields the "doc" field wasn't getting serialized properly, I fixed that in this PR. Running mzcompose to check this out (taking a while).

**Comment by moulimukherjee** [comment]
@def- Oh I think it's the envelope `DEBEZIUM`. I could reproduce it in testdrive. Thanks for catching this! Will put up a fix.

**Comment by moulimukherjee** [comment]
@def- It should be fixed now.

**Review by def- — APPROVED** [review]
Thanks for waiting for my tests. I postponed the Python-avro-based check until we reach a conclusion on how we want to test it.

File: misc/python/materialize/checks/sink.py
Comment (def-) [inline misc/python/materialize/checks/sink.py:707]: Is it expected to have the comment on the row instead of on the envelope?

Comment (benesch) [reply misc/python/materialize/checks/sink.py:707]: Yes, see discussion here: https://materializeinc.slack.com/archives/CUFU852KT/p1696642943413989

**Comment by umanwizard** [comment]
Dennis Felsing ***@***.***> writes:

> @def- commented on this pull request.
>
>
>
>>  
>                  $ schema-registry-verify schema-type=avro subject=sink-sink-comments2-value
> -                {"type":"record","name":"envelope","doc":"comment on view sink_source_comments_view","fields":[{"name":"before","type":["null",{"type":"record","name":"row","fields":[{"name":"l_k","type":"string"},{"name":"l_v1","type":["null","string"],"default":null,"doc":"doc on l_v1"},{"name":"l_v2","type":["null","long"],"default":null,"doc":"value doc on l_v1"},{"name":"c","type":"long"}]}],"default":null},{"name":"after","type":["null","row"],"default":null}]}
> +                {"type":"record","name":"envelope","fields":[{"name":"before","type":["null",{"type":"record","name":"row","doc":"comment on view sink_source_comments_view","fields":[{"name":"l_k","type":"string"},{"name":"l_v1","type":["null","string"],"default":null,"doc":"doc on l_v1"},{"name":"l_v2","type":["null","long"],"default":null,"doc":"value doc on l_v2"},{"name":"c","type":"long"}]}],"default":null},{"name":"after","type":["null","row"],"default":null}]}
>
> Is it expected to have the comment on the row instead of on the
> envelope?

Yes. Discussed in Slack:
https://materializeinc.slack.com/archives/CUFU852KT/p1696642943413989

File: test/testdrive/kafka-avro-sinks-doc-comments.td
Comment (benesch) [reply test/testdrive/kafka-avro-sinks-doc-comments.td:13]: > This does not prevent the user from explicitly adding a DOC ON options in their create sink sql though.

We should add a separate feature flag for the use of `DOC ON` comments, so that we can properly follow the feature lifecycle for this feature. It should start out in private preview while we're validating its design and implementation with Pluralsight. That way we can make breaking changes if necessary.

File: doc/developer/design/20230903_avro_doc.md
Comment (benesch) [reply doc/developer/design/20230903_avro_doc.md:117]: 👍🏽

**Review by benesch — COMMENTED** [review]
I only had time to quickly skim, but looks really solid. Basically no substantive comments. 👌🏽

**Review by ParkMyCar — APPROVED** [review]

File: test/testdrive/kafka-avro-sinks-doc-comments.td
Comment (moulimukherjee) [reply test/testdrive/kafka-avro-sinks-doc-comments.td:13]: Sounds good. Let me add that.

Comment (moulimukherjee) [reply test/testdrive/kafka-avro-sinks-doc-comments.td:13]: Done! Put it behind `enable_sink_doc_on_option`

File: misc/python/materialize/mzcompose/__init__.py
Comment (def-) [inline misc/python/materialize/mzcompose/__init__.py:72]: should also add `"enable_sink_doc_on_option": "true"` here so we can use it easily in testing.

**Review by def- — CHANGES_REQUESTED** [review]

File: misc/python/materialize/mzcompose/__init__.py
Comment (moulimukherjee) [reply misc/python/materialize/mzcompose/__init__.py:72]: done!

**Comment by moulimukherjee** [comment]
All the comments are addressed, enabling auto-merge.

**Pull request merged**
